### ISSUE TITLE
[no-dynamic-delete] Compatibility with typescript's noPropertyAccessFromIndexSignature 

### ISSUE THREAD

**Issue by iliubinskii** [typescript-eslint/typescript-eslint#3504]
**Repro**

Example from no-dynamic-delete rule documentation:
```TS
// Can be replaced with the constant equivalents, such as container.aaa
delete container['aaa'];
```

**Expected Result**

Typescript's noPropertyAccessFromIndexSignature may force you to change container.aaa to container['aaa'], while this rule will force you to change it back to container.aaa.

So, I would expect the above code to pass because:
1. 'aaa' is NOT dynamic, it is fixed.
2. You may not always change it to container.aaa.

**Actual Result**

ESLint warning issued.

**Additional Info**

-
**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.26.0` |
| `@typescript-eslint/parser`        | `4.26.0` |
| `TypeScript`                       | `4.3.2` |
| `ESLint`                           | `7.28.0` |
| `node`                             | `14.17.0` |

**Comment by bradzacher** [comment]
could you flesh out your example more?
`delete container['aaa']` is ambiguous in relation to `noPropertyAccessFromIndexSignature`.
Please add types and make it a minimal repro.

**Comment by ilyub** [comment]
```ts
const container: Record<string, string> = {};

delete container["aaa"];
```

I get this error:
![1](https://user-images.githubusercontent.com/70687432/121306343-63211a80-c907-11eb-8827-7ed8637cd34b.jpg)

no-dynamic-delete rule has fixer that changes the above code to:
```ts
const container: Record<string, string> = {};

delete container.aaa;
```

Then I get this error:
![2](https://user-images.githubusercontent.com/70687432/121306672-ce6aec80-c907-11eb-9c7a-f3d6019cd030.jpg)

Typescript's noPropertyAccessFromIndexSignature rule has its own fixer that changes it back to container["aaa"].

But the purpose of no-dynamic-delete rule, as can be read in the documentation, is the following:
"Deleting dynamically computed keys can be dangerous and in some cases not well optimized"

"aaa" is not dynamically computed.

So, if no-dynamic-delete rule would abide to its purpose then it would not conflict with other rules.

There other rules in TypeScript and ESLint that control use of container.aaa and container["aaa"].

These other rules do it better than no-dynamic-delete.

**Comment by bradzacher** [comment]
In order to detect this, we would need to use type information to determine what your tsconfig is using.
this is a breaking change.

TBH it's not really a good change for the rule because it'd be the only type information being used.
But we can add it to the backlog.

**Comment by ilyub** [comment]
To me the rule just needs to allow the use of constants: container["aaa"], container['aaa']
and forbid the use of variables (e.g. container[variable]) or anything that contains computation (e.g. container["a" + "a" + "a"]).
This will fix an obvious discrepancy when the rule tells that container["aaa"] is dynamically computed when it is not.
Allowing container["aaa"] pattern can always be added as additional option.

**Comment by dipunm** [comment]
+1 for this. Ran into this today.

It could be a configuration option like https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/dot-notation.md does.

I think we can detect what is passed into the accessor and if it is a constant, we can (optionally based on the configuration?) allow the code.
